### PR TITLE
Assignment6 - Alexa Darrington and Johanne McClenahan

### PR DESCRIPTION
**Pull request by JohanneJayde**
Alexa's Github - @alyxadarr
Johanne's Github - @JohanneJayde

### CODE REVIEW

**Comment by Purplepigs0129** [comment]
Reviewing! :)

File: Calculate/Program.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿using System.Globalization;
+using ConsoleUtilites;
+namespace Calculate;
+
+public class Program : ProgramBase
+{
```
Comment (Purplepigs0129) [inline Calculate/Program.cs:6]: You are missing the Writeline and Readline properties they also should be setup as init-only setters. They should also be = to the console versions of them. This is what we have for them as its hard to put everything in one comment but you do need these.
 public Action<string> WriteLine { get; init; } = Console.WriteLine;
    public Func<string> ReadLine { get; init; } = Console.ReadLine;

Comment (Purplepigs0129) [reply Calculate/Program.cs:6]: actually, now that I see that is is extending a base class I am assuming this was for the EC attempt but its not here so maybe it accidentally got deleted and that's where they are set up

Comment (JohanneJayde) [reply Calculate/Program.cs:6]: This is for EC. The redirection portion was there originally but then was moved to ConsoleUtilities assembly. I originally included the class library project with submission but omitted it. I put it back so it's clearer on how we did it.

File: Calculate/Calculator.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    public IReadOnlyDictionary<char, Func<int,int,int>> MathematicalOperations { get; } = new Dictionary<char, Func<int, int, int>>()
+        { 
+        {'+',Add},
+        {'-',Subtract},
+        {'*',Multiply },
+        {'/', Divide}
+        };
+    
+    public static int Add(int num1, int num2)
```
Comment (Purplepigs0129) [inline Calculate/Calculator.cs:15]: For all of the methods, I know that it is much simpler to do it this way but the project specifically says that you take in 2 params and return a third. You have the 2 params but no third should be as easy as int answer = num1 operation num2; return answer;

File: Calculate/Calculator.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        return num1 + num2;
+    }
+    public static int Subtract(int num1, int num2)
+    {
+        return num1 - num2;
+
+    }
+    public static int Multiply(int num1, int num2)
```
Comment (Purplepigs0129) [inline Calculate/Calculator.cs:24]: Should be named multiple not multiply even though that is what we are doing here

File: Calculate.Tests/CalculateTests.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+using System.Linq.Expressions;
+using Xunit;
+
+namespace Calculate.Tests;
+
+    public class CalculateTests
+    {
+        public Calculator Calculator { get;}
+
+        public CalculateTests()
```
Comment (Purplepigs0129) [inline Calculate.Tests/CalculateTests.cs:11]: Definitely need a constructor for our tests great work! :)

File: Calculate.Tests/ProgramTests.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        InitProgram.WriteLine("Testing File Writing");
+        InitProgram.WriteLine("Testing File Writing 2");
+
+        using StreamReader streamReader = new(TempPath);
+        Assert.Equal("Testing File Writing", streamReader.ReadLine());
+        Assert.Equal("Testing File Writing 2", streamReader.ReadLine());
+
+    }
+    [Fact]
+    public void Constructor_InitFileReadLine_ExpectedBehavior()
```
Comment (Purplepigs0129) [inline Calculate.Tests/ProgramTests.cs:74]: Assuming this is EC for file intake Great Work!

**Review by Purplepigs0129 — COMMENTED** [review]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ❌
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ❌
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ❌
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔
Extra Credit
Do one of the following two options (or both if you want extra, extra credit) :)

Refactor the redirect portion of the Program class into 'ProgramBase`
Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Great Work! Somehow you guys are missing the props even though I bet they are in the ProgramBase class only problem is that is not currently here. Make sure to have the file back or add the props to Program so that it works properly. Otherwise just a few other small fixes.

File: Calculate/Calculator.cs (hunk 3 above)
Comment (JohanneJayde) [reply Calculate/Calculator.cs:24]: I don't think it matters as it's just the method caller. Although it's on the requirements, it doesn't affect the code and means the same thing

File: Calculate.Tests/CalculateTests.cs (hunk 4 above)
Comment (JohanneJayde) [reply Calculate.Tests/CalculateTests.cs:11]: Thank!

File: Calculate.Tests/ProgramTests.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+        streamWriter.Close();
+
+        Assert.Equal("Testing File Reading", InitProgram.ReadLine());
+        Assert.Equal("Testing File Reading", InitProgram.ReadLine());
+
+    }
+
+    [Fact]
+    public void Constructor_DefaultReadFileWrite()
```
[the remark this answers is not in the document]
Comment (JohanneJayde) [reply Calculate.Tests/ProgramTests.cs:88]: I will delete this test. Somehow build didn't have an issue with it. Thanks for finding it :)

File: Calculate.Tests/ProgramTests.cs (hunk 5 above)
Comment (JohanneJayde) [reply Calculate.Tests/ProgramTests.cs:74]: Thank you :)

Comment (JohanneJayde) [reply Calculate.Tests/ProgramTests.cs:74]: No it's not. This is for testing when we initialize WriteLine and ReadLine properties in program.

File: Calculate/Calculator.cs (hunk 2 above)
Comment (JohanneJayde) [reply Calculate/Calculator.cs:15]: I don't think this is right. Your solution only adds a local variable and returns that. My interpretation of this requirement was that the it was supposed to mean Fun<int,int,int> was fulfilling it. As all methods have two parameters as inputs and an output parameter

**Comment by Joshua-Pedersen** [comment]
will review

File: Calculate/Calculator.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        return num1 * num2;
+    }
+    public static int Divide(int num1, int num2)
```
Comment (Joshua-Pedersen) [inline Calculate/Calculator.cs:28]: Divide by zero is possible with this implementation and throws an unhandled error.

File: Calculate.Tests/ProgramTests.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
+using Xunit;
+namespace Calculate.Tests;
+
+public class ProgramTests : IDisposable
+{
+
+    // Will never be null as properties is set in constructor
+    #pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
```
Comment (Joshua-Pedersen) [inline Calculate.Tests/ProgramTests.cs:8]: Nice job, explaining why the warning was suppressed

File: Calculate.Tests/CalculateTests.cs
Comment (Joshua-Pedersen) [inline Calculate.Tests/CalculateTests.cs:1]: Worth adding a test to ensure division by 0 is handled

File: Calculate.Tests/CalculateTests.cs [hunk 9]
```diff
@@ -0,0 +1,77 @@
+﻿
+using System.Linq.Expressions;
```
Comment (Joshua-Pedersen) [inline Calculate.Tests/CalculateTests.cs:2]: Non-needed using statement

**Review by Joshua-Pedersen — COMMENTED** [review]
## Instructions

- Create a *Console* project called "Calculate.". ✔
- Define a Program Class
  - Define two init-only setter properties, `WriteLine` and `ReadLine`, that contain delegates for writing a line of text and reading a line of text respectively ✔
  - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
  - Set the default behavior for the `WriteLine` and `ReadLine` properties to invoke `System.Console` versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
  - Define static `Add`, `Subtract`, `Multiple`, and `Divide` methods that have two parameters and return a third parameter. ✔
  - Define a read-only property, `MathematicalOperations`, of type `System.Collections.Generics.IReadOnlyDictionary<TKey,TValue>` that:
    - is initialized to a `System.Collections.Generics.Dictionary<<TKey,TValue>` instance that. ✔
      - Uses `char` for the key corresponding to the operators +, -, *, and /. ✔
      - Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
  - Implement a `TryCalculate` method following "TryParse" pattern ✔
    - Valid `calculation` expressions include such strings as "3 + 4", "42 - 2", etc. ✔
    - If there is no whitespace around the operator, you can assume the `calculation` is invalid and return false. Similarly if the operands are not integers. ✔
    - Use `string.Split()`, pattern matching, logical and operators to parse the string in their entirety ✔
    - Index into the `MathematicalOperations` method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the `WriteLine`/`ReadLine` properties on `Program` for testing the input and output of your program. ✔

## Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

- Refactor the redirect portion of the `Program` class into 'ProgramBase`
- Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
- Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- Place all shared project items into a `Directory.Build.targets` file.
- nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ❌ Add a test for divide by zero
- Choose simplicity over complexity ✔

File: Calculate.Tests/ProgramTests.cs (hunk 8 above)
Comment (JohanneJayde) [reply Calculate.Tests/ProgramTests.cs:8]: Thank you !

File: Calculate.Tests/CalculateTests.cs
Comment (JohanneJayde) [reply Calculate.Tests/CalculateTests.cs:1]: I agree. I saw many people do this and it totally spaced my mind. I will add it.

File: Calculate/Calculator.cs (hunk 7 above)
Comment (JohanneJayde) [reply Calculate/Calculator.cs:28]: Yes, will fix soon

File: Calculate/Calculator.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
+﻿using System.Globalization;
+using System.Numerics;
+
+namespace Calculate;
+
+public class Calculator<T> where T : INumber<T>
```
Comment (JohanneJayde) [inline Calculate/Calculator.cs:6]: For EC, the `Add`, `Subtract`, `Multiply`, and `Divide` methods have been made non-static. This requirement was fulfilled before.

Making operations generic would force us to make `Calculator` generic due to `MathematicalOperations` needing to hold the correct type for each operator. Having a generic `Calculator` with static operations will cause `CA1000`. One of the suggestions for fixing this was to make the static members into instance members. 

We did this because if someone creates a `Calculator` of type `double` then all operations would need to be of type double. So making them instance-based ties them to whichever type `Calculator` has been configured to operate on.

File: Calculate/Calculator.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        MathematicalOperations = new Dictionary<char, Func<T, T, T>>()
+        {
+            {'+',Add},
+            {'-',Subtract},
+            {'*',Multiply},
+            {'/',Divide }
+        };
+    }
+
+    public T Add(T value, T value2)
```
Comment (JohanneJayde) [inline Calculate/Calculator.cs:21]: Should the operation methods be better be private? If we are assuming that a user will always use `TryCalcalulate` then we would want to make it so they can't call these methods.

File: Calculate/Calculator.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        return value * value2;
+    }
+
+    public T Divide(T value, T value2)
+    {
+        if (value2 == T.Zero)
+            throw new DivideByZeroException("Cannot divide by zero");
+
+        return value / value2;
+    }
```
Comment (JohanneJayde) [inline Calculate/Calculator.cs:42]: These were `static` before doing EC for `MathematicalOperations` The reasoning behind making them instance-based was due to `CA1000`.

`Calculator` has to be updated to be generic to store the generic `Add`, `Subtract`, `Divide`, and `Multiply`. This causes 'CA1000`. Based on the documentation, it would be better to make operations instance-based so that they are tied to a `Calculator`'s given type. 

However, I do understand that since we are using Generic Maths it could be different and the disabling warning is completely valid.

`TryCalculate` has also been made generic.

File: Calculate/Calculator.cs [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+    public T Multiply(T value, T value2)
+    {
+        return value * value2;
+    }
+
+    public T Divide(T value, T value2)
+    {
+        if (value2 == T.Zero)
+            throw new DivideByZeroException("Cannot divide by zero");
```
Comment (BenjaminMichaelis) [inline Calculate/Calculator.cs:39]: Divide by zero by default throws a `System.DivideByZeroException` in the BCL, this is redundant and unecessary

File: Calculate/Calculator.cs (hunk 10 above)
Comment (BenjaminMichaelis) [inline Calculate/Calculator.cs:6]: Consider taking in two type parameters in order to return a different precision than what is equated with (ex: int / int = double)

File: Calculate.Tests/CalculateTests.cs [hunk 14]
```diff
@@ -0,0 +1,88 @@
+﻿using System.Linq.Expressions;
+using Xunit;
+
+namespace Calculate.Tests;
+
+    public class CalculateTests
+    {
+        public Calculator<double> Calculator { get;}
```
Comment (BenjaminMichaelis) [inline Calculate.Tests/CalculateTests.cs:8]: Since this isn't being used for all tests, I would put this in the test itself

File: Calculate.Tests/CalculateTests.cs [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        [Fact]
+        public void Add_ValidNumbers_ReturnsExpectedResult()
+        {
+            Assert.Equal(6, Calculator.Add(3.0, 3));
+        }
+        [Fact]
+        public void Subtract_ValidNumbers_ReturnsExpectedResult()
+        {
+            Calculator<float> floatCalculator = new();
+            Assert.Equal(6.4 - 3.7, Calculator.Subtract(6.4, 3.7));
```
Comment (BenjaminMichaelis) [inline Calculate.Tests/CalculateTests.cs:24]: floatCalculator doesn't actually get used her
